src/mca.jl

```diff
@@ -0,0 +1,354 @@
+
+    # Get the object factor scores (F) and variable factor scores (G).
+    P, D, Q = svd(ca.SR)
+    Dq = Diagonal(D)[:, 1:d]
+
+    # Check that there are no repeated non-zero eigenvalues.
+    d = diff(D[D .> 1e-10])
+    if maximum(d) >= -1e-10
+        @warn("The indicator matrix has repeated non-zero eigenvalues")
+    end
+
+    Wr = Diagonal(sqrt.(ca.rm))
+    Wc = Diagonal(sqrt.(ca.cm))
+    ca.F = Wr \ P * Dq
+    ca.G = Wc \ Q * Dq
+
+    # Get the eigenvalues
+    ca.I = D .^ 2
+end
+
+function fit(::Type{CA}; X::AbstractMatrix, d::Int = 5)
+    ca = CA(X)
+    fit!(ca, d)
+    return ca::CA
+end
+
+"""
+    ca
+
+Fit a correspondence analysis using the data array `X` whose rows are
+the objects and columns are the variables.  The first `d` components
+are retained.
+"""
+function ca(X, d)
+    c = fit(CA, X, d)
+    return c
+end
+
+objectscores(ca::CA) = ca.F
+variablescores(ca::CA) = ca.G
+inertia(ca::CA) = ca.I
+
+"""
+Multiple Correspondence Analysis
+"""
+mutable struct MCA{T<:Real} <: LinearDimensionalityReduction
+
+    # The underlying corresponence analysis
+    C::CA{T}
+
+    # Variable names
+    vnames::Vector{String}
+
+    # Map values to integer positions
+    rd::Vector{Dict}
+
+    # Map integer positions to values
+    dr::Vector{Dict}
+
+    # Split the variable scores into separate arrays for
+    # each variable.
+    Gv::Vector{Matrix{T}}
+
+    # Number of nominal variables
+    K::Int
+
+    # Total number of categories in all variables
+    J::Int
+
+    # Eigenvalues
+    unadjusted_eigs::Vector{Float64}
+    benzecri_eigs::Vector{Float64}
+    greenacre_eigs::Vector{Float64}
+end
+
+objectscores(mca::MCA) = mca.C.F
+variablescores(mca::MCA) = mca.Gv
+
+# Split the variable scores to a separate array for each
+# variable.
+function xsplit(G, rd)
+    K = [length(di) for di in rd]
+    Js = cumsum(K)
+    Js = vcat(1, 1 .+ Js)
+    Gv = Vector{Matrix{eltype(G)}}()
+    for j in eachindex(K)
+        g = G[Js[j]:Js[j+1]-1, :]
+        push!(Gv, g)
+    end
+    return Gv
+end
+
+# Calculate the eigenvalues with different debiasings.
+function get_eigs(I, K, J)
+    ben = zeros(length(I))
+    gra = zeros(length(I))
+    Ki = 1 / K
+    f = K / (K - 1)
+    for i in eachindex(I)
+        if I[i] > Ki
+            ben[i] = (f * (I[i] - Ki))^2
+        end
+    end
+
+    unadjusted = I ./ sum(I)
+    gt = f * (sum(abs2, I) - (J - K) / K^2)
+
+    return unadjusted, ben ./ sum(ben), ben ./ gt
+end
+
+# constructor
+
+function MCA(Z; vnames = [])
+
+    if length(vnames) == 0
+        vnames = ["v$(j)" for j = 1:size(Z, 2)]
+    end
+
+    # Get the indicator matrix
+    X, rd, dr = make_indicators(Z)
+
+    # Create the underlying correspondence analysis value
+    C = CA(X)
+
+    # Number of nominal variables
+    K = size(Z, 2)
+
+    # Total number of categories in all variables
+    J = size(X, 2)
+
+    return MCA(C, vnames, rd, dr, Matrix{Float64}[], K, J, zeros(0), zeros(0), zeros(0))
+end
```

Comment:
Move the code into the `fit` function. There is no need for a special model constructor.